### PR TITLE
stats: add iterate methods for Scope, and a lazy find-from-string implementation.

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: Adds iterate() functionality for stats scopes. This makes it possible to robustly find a stat by string in prod code, even if the caller does not know whether some segments of the name were created via StatNameDynamicStorage, so a new Stats::Utility API is added to encapsulate the lookup and caching of stats by string.

Note that the existing deprecated counterFromStatName only works for fully symbolic names.
Additional Description: This can potentially enable a hierarchal stats exploration admin interface (though we'd need to be able to iterate over sub-scopes too).
Risk Level: low -- these new capabilities are tested, but not called from prod code yet. They may prove useful at Google and in Nighthawk, per the referenced bug.
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a
Fixes: #11799

### CODE REVIEW

**Comment by jmarantz** [comment]
@eziskind @oschaaf FYI

File: source/common/stats/utility.h [hunk 1]
```diff
@@ -206,5 +207,55 @@ class Utility {
                                                StatNameTagVectorOptConstRef tags = absl::nullopt);
 };
 
+/**
+ * Holds a reference to a stat by name. Note that the stat may not be created
+ * yet at the tome CachedReference is created. Calling find() then does a lazy
```
Comment (mattklein123) [inline source/common/stats/utility.h:212]:
```suggestion
 * yet at the time CachedReference is created. Calling find() then does a lazy
```

File: source/common/stats/utility.h [hunk 2]
```diff
@@ -206,5 +207,55 @@ class Utility {
...
+    });
+    if (stat == nullptr) {
+      return absl::nullopt;
+    }
+    return *stat;
+  }
+
+private:
+  Scope& scope_;
+  std::string name_;
```
Comment (mattklein123) [inline source/common/stats/utility.h:256]: nit: const

File: source/common/stats/utility.h [hunk 3]
```diff
@@ -206,5 +207,55 @@ class Utility {
...
+ */
+template <class StatType> class CachedReference {
+public:
+  CachedReference(Scope& scope, absl::string_view name) : scope_(scope), name_(std::string(name)) {}
+
+  /**
+   * Finds the named stat, if it exists, returning it as an optional.
+   */
+  absl::optional<std::reference_wrapper<StatType>> find() {
+    StatType* stat = stat_.get([this]() -> StatType* {
```
Comment (mattklein123) [inline source/common/stats/utility.h:235]: Seems like find() should really only be called once. Can we assert somehow that `stat_` has never been set?

Comment (jmarantz) [reply source/common/stats/utility.h:235]: The intended semantics here are that `find()` does caching itself so it's fine to call it lots of times.

I'm not clear on what you are suggesting as an alternative?

Comment (mattklein123) [reply source/common/stats/utility.h:235]: I just meant that after you call find() once, the result is cached. Sorry I think I didn't give a full comment. Should there be some different between initialize() and find() in that initialize() is called once and find() should be called after initialize()? I don't feel strongly about it. In that case you could s/find/get.

File: source/common/stats/utility.h (hunk 1 above)
Comment (jmarantz) [reply source/common/stats/utility.h]: done

File: source/common/stats/utility.h (hunk 3 above)
Comment (jmarantz) [reply source/common/stats/utility.h:235]: changed to `get()`

File: source/common/stats/utility.h (hunk 2 above)
Comment (jmarantz) [reply source/common/stats/utility.h]: done

**Comment by jmarantz** [comment]
I think the messaging is reasonably clear about the slowness of the initial lookup in the doc for `CachedReference`. If you have any suggestions on how to improve the doc, would love to hear them.

The intended usage is during initialization when capturing a reference to stat which we only know by name.

**Pull request merged**